src/Adapter/AbstractObjectModelRepository.php

```diff
@@ -66,35 +72,76 @@ protected function assertObjectModelExists(int $id, string $objectTableName, str
      * @param string $objectModelClass
      * @param string $exceptionClass
+     * @param ShopId|null $shopId
      *
      * @return ObjectModel
      *
      * @throws CoreException
      */
-    protected function getObjectModel(int $id, string $objectModelClass, string $exceptionClass): ObjectModel
+    protected function getObjectModel(int $id, string $objectModelClass, string $exceptionClass, ?ShopId $shopId = null): ObjectModel
     {
-        try {
-            $objectModel = new $objectModelClass($id);
+        return $this->fetchObjectModel($id, $objectModelClass, $exceptionClass, null);
+    }
 
-            if ((int) $objectModel->id !== $id) {
-                throw new $exceptionClass(sprintf('%s #%d was not found', $objectModelClass, $id));
-            }
-        } catch (PrestaShopException $e) {
-            throw new CoreException(
-                sprintf(
-                    'Error occurred when trying to get %s #%d [%s]',
-                    $objectModelClass,
-                    $id,
-                    $e->getMessage()
-                ),
-                0,
-                $e
-            );
-        }
+    /**
+     * @param int $id
+     * @param string $objectModelClass
+     * @param string $exceptionClass
+     *
+     * @return ObjectModel
+     *
+     * @throws CoreException
+     */
+    protected function getObjectModelForShop(int $id, string $objectModelClass, string $exceptionClass, ShopId $shopId): ObjectModel
+    {
+        $objectModel = $this->fetchObjectModel($id, $objectModelClass, $exceptionClass, $shopId->getValue());
+
+        // The object is fetched before checking the association, so that the NotFoundException has the priority over the NoAssociationException
+        $this->checkShopAssociation($id, $objectModelClass, $shopId);
+
+        // Force id_shop_list right away so that DB modification use the appropriate shop and not the one from context
+        $objectModel->id_shop_list = [$shopId->getValue()];
 
         return $objectModel;
     }
 
+    /**
+     * @param int $id
+     * @param string $objectModelClassName
+     * @param ShopId $shopId
+     *
+     * @throws ShopAssociationNotFound
+     */
+    protected function checkShopAssociation(int $id, string $objectModelClassName, ShopId $shopId): void
+    {
+        $modelDefinition = $objectModelClassName::$definition;
+        $objectTable = $modelDefinition['table'];
+        $primaryColumn = $modelDefinition['primary'];
+
+        $query = new DbQuery();
+        $query
+            ->select('e.`' . $primaryColumn . '` as id')
+            ->from($objectTable . '_shop', 'e')
+            ->where('e.`' . $primaryColumn . '` = ' . $id)
+            ->where('e.`id_shop` = ' . $shopId->getValue())
+        ;
+
+        try {
+            $row = Db::getInstance()->getRow($query, false);
+        } catch (PrestaShopDatabaseException $e) {
+            $row = false;
+        }
+
+        if (!isset($row['id'])) {
```

Comment:
```suggestion
        if (empty($row['id'])) {
```
Should we really need to find zero values?